### PR TITLE
CC: What Every Computer Scientist Should Know About Floating-Point Arithmetic (#950)

### LINKED ISSUE #950: What Every Computer Scientist Should Know About Floating-Point Arithmetic
author: @bartlettroscoe 
eb member: @pagrubel 

Seems like a very thorough coverage of the issues...

https://docs.oracle.com/cd/E19957-01/806-3568/ncg_goldberg.html

### PR DESCRIPTION
**Pull request by bartlettroscoe**
# Description

EB Member: @pagrubel 

Resolves #950 

CC article for "What Every Computer Scientist Should Know About Floating-Point Arithmetic" with some summary and modern commentary.

## PR checklist for files displayed on bssw.io site

* [x] ***[Author]*** `@mention` the BSSw.io editorial board member `@<eb-member-id>` in **Description** above assigned to shepherd your PR.
* [x] ***[Author]*** Add the `<issue-id>` in the **Description** above for the associated GitHub Issue.
* ~~[ ] ***[Author]*** Ensure `wikize_refs.py -i <base>.md` is run and commit (if using [wikize_refs.py])~~.
* [x] ***[Author]*** Inspect the content in the `*.md` file(s) as rendered in GitHub for this PR.
* [x] ***[EB Mem]*** Assign this PR to the EB member `<eb-member-id>`.
* [x] ***[EB Mem]*** Assign this PR to the author of the PR `<pr-author-id>`.
* [x] ***[EB Mem]*** Add one or more Reviewers.
* [x] ***[EB Mem]*** Add label `content: <content-type>` for the type of contribution.
* [x] ***[EB Mem]*** Add to Project `Content Development` (see [Content Development]).
* [x] ***[EB Mem]*** Add [meta-data] to the `*.md` file(s) (set `Publish: yes`).
* [x] ***[EB Mem]*** Add label `preview` (so PR branch will be merged to 'preview' branch and watch for possible merge failures).
* [x] ***[EB Mem]*** Rebuild [preview] site and confirm new content is there, renders correctly and is returned in searches.
* [x] ***[Author]*** Make any final changes to the PR based on feedback.
* ~~[ ] ***[Author]*** Ensure `wikize_refs.py -i <base>.md` is run and commit (if using [wikize_refs.py]).~~
* [x] ***[EB Mem]*** Rebuild [preview] site and re-confirm content looks correct.
* [x] ***[EB Mem]*** Ensure at least one reviewer signs off on the final changes.
* [x] ***[EB Mem]*** Move the PR to "Ready to Publish" in [Content Development].
* [x] ***[EB Mem]*** Leave comment and @mention `@betterscientificsoftware/bssw-maint` (**BSSw Maint**) asking to carry out final publication steps.
* [x] ***[BSSw Maint]*** Verify that all needed files are present in the PR (article, images, updates to Site/Homepage.md carousel and/or Site/Announcements/Announcements.md as appropriate).
* [x] ***[BSSw Maint]*** Merge PR. (Should automatically move to "Done" in [Content Development].)
* [ ] ***[BSSw Maint]*** Verify new contribution shows up on [bssw.io] as expected.

NOTE:
* Checklist items prefixed with ***[Author]*** are expected to be performed by the author of the PR or can be performed by the author.
* Checklist items prefixed with ***[EB Mem]*** must be performed by a [BSSw.io Editorial Board (EB) Member](https://betterscientificsoftware.github.io/bssw.io/bssw_members.html).
* Checklist items prefixed with ***[BSSw Maint]*** must be performed by someone in the `@betterscientificsoftware/bssw-maint` team (hint: type `@`, `b`, `s`, `s`, `w`, `-`, `m`  to auto-complete to `@betterscientificsoftware/bssw-maint`).

[preview]: https://preview.bssw.io
[bssw.io]: https://bssw.io
[Content Development]: https://github.com/betterscientificsoftware/bssw.io/projects/3
[BSSw Internal]: https://github.com/betterscientificsoftware/bssw.io/projects/2
[meta-data]: https://betterscientificsoftware.github.io/bssw.io/bssw_styling_common.html#metadata-section
[wikize_refs.py]: https://github.com/betterscientificsoftware/bssw.io/blob/master/utils/README.md#wikize_refspy

### CODE REVIEW

**Comment by bartlettroscoe** [comment]
I manually fired off the rebuilding of the preview.bssw.io site and the article looks good:

* https://preview.bssw.io/items/what-every-computer-scientist-should-know-about-floating-point-arithmetic

There is not really any fancy formatting going on other than perhaps a couple of Greek symbols &epsilon; and  &infin; which are also displayed correctly.

File: CuratedContent/WhatEveryComputerScientistShouldKnowAboutFloatingPointArithmetic.md [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+* bit-wise floating-point number data-structures (i.e. significant digits, exponents, sign),
+* rounding and accuracy requirements for basic floating-point operations,
+* overflow, underflow and special quantities (i.e. (signed) zero, demormalized numbers, NaNs, (signed) infinity &infin;),
+* floating-point exceptions, flags, and trap handlers,
+* rounding modes, and
+* conversions with integers and decimal (i.e. human-readable text-based) formats.
+
+Great attention is given to the issue of catastrophic cancellation (the subtraction of two nearby numbers where the operands are also subject to rounding errors) and related floating-point issues.
+Several examples are given of how different numerical expressions can be reformulated to eliminate a catastrophic cancellation and/or avoid overflow/overflow and thereby increase the robustness and accuracy of the floating-point calculation to a wide range of input data.
```
Comment (pagrubel) [inline CuratedContent/WhatEveryComputerScientistShouldKnowAboutFloatingPointArithmetic.md:47]: overflow/underflow

File: CuratedContent/WhatEveryComputerScientistShouldKnowAboutFloatingPointArithmetic.md [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+Several compilers and systems over the years have been known to violate this and other necessary basic foundations of floating-point and the IEEE 754 standard, especially when compiler optimizations are turned up.
+In addition, full implementation of the IEEE 754 standard on a system is rare and controlling behavior such as floating-point exceptions, rounding modes, and trap handlers is difficult or impossible in a portable way with many programming languages and libraries.
+This is especially true on some of the more bleeding-edge high-performance computing (HPC) accelerators and systems.
+As a result, numerical algorithm developers often have to rely on a smaller subset of the IEEE standard in order to develop a maximally portable robust implementation.
+
+Finally, the paper contains quite a few proofs of floating-point arithmetic and these proofs serve several purposes.
+First, the proofs establish the foundation of floating-point operations which demonstrate that floating-point is not black magic but is a well established mathematical approximation/discritization of real number arithmetic.
```
Comment (pagrubel) [inline CuratedContent/WhatEveryComputerScientistShouldKnowAboutFloatingPointArithmetic.md:61]: discretization

File: CuratedContent/WhatEveryComputerScientistShouldKnowAboutFloatingPointArithmetic.md [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+* floating-point formats,
+* rounding error (e.g. guard digits) and measures of error (e.g. machine epsilon &epsilon;, absolute error),
+* loss of basic associativity and loss of other basic properties of real arithmetic, and
+* catastrophic cancellation.
+
+In addition, the author covers IEEE floating-point standards including the IEEE 754 binary standard which addresses:
+
+* bit-wise floating-point number data-structures (i.e. significant digits, exponents, sign),
+* rounding and accuracy requirements for basic floating-point operations,
+* overflow, underflow and special quantities (i.e. (signed) zero, demormalized numbers, NaNs, (signed) infinity &infin;),
```
Comment (pagrubel) [inline CuratedContent/WhatEveryComputerScientistShouldKnowAboutFloatingPointArithmetic.md:41]: is it denormalize

File: CuratedContent/WhatEveryComputerScientistShouldKnowAboutFloatingPointArithmetic.md [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+<!---
+Publish: yes 
+Pinned: no
+RSS update: ???
+Topics: Programming languages, High-performance computing (HPC), Reproducibility, Refactoring
+--->
+
+<!---
+LocalWords:  associativity demormalized
```
Comment (pagrubel) [inline CuratedContent/WhatEveryComputerScientistShouldKnowAboutFloatingPointArithmetic.md:78]: should this be denormalize

**Review by pagrubel — CHANGES_REQUESTED** [review]
@bartlettroscoe This article looks great. I marked some very minor changes to be made.

File: CuratedContent/WhatEveryComputerScientistShouldKnowAboutFloatingPointArithmetic.md (hunk 1 above)
Comment (bartlettroscoe) [reply CuratedContent/WhatEveryComputerScientistShouldKnowAboutFloatingPointArithmetic.md]: Thanks!  Addressed in commit bcd4e2f264bb664f80eeaa4be26fbdde6918b46b.

File: CuratedContent/WhatEveryComputerScientistShouldKnowAboutFloatingPointArithmetic.md (hunk 2 above)
Comment (bartlettroscoe) [reply CuratedContent/WhatEveryComputerScientistShouldKnowAboutFloatingPointArithmetic.md]: Thanks!  Addressed in commit bcd4e2f264bb664f80eeaa4be26fbdde6918b46b.

File: CuratedContent/WhatEveryComputerScientistShouldKnowAboutFloatingPointArithmetic.md (hunk 3 above)
Comment (bartlettroscoe) [reply CuratedContent/WhatEveryComputerScientistShouldKnowAboutFloatingPointArithmetic.md:41]: Thanks!  Addressed in commit bcd4e2f264bb664f80eeaa4be26fbdde6918b46b.

File: CuratedContent/WhatEveryComputerScientistShouldKnowAboutFloatingPointArithmetic.md (hunk 4 above)
Comment (bartlettroscoe) [reply CuratedContent/WhatEveryComputerScientistShouldKnowAboutFloatingPointArithmetic.md]: Thanks!  Addressed in commit bcd4e2f264bb664f80eeaa4be26fbdde6918b46b.

**Comment by bartlettroscoe** [comment]
@pagrubel, thanks for the fast review.  All of the issues you flagged should be addressed in commit bcd4e2f264bb664f80eeaa4be26fbdde6918b46b that I just pushed.

**Comment by bartlettroscoe** [comment]
> @bartlettroscoe you did not change denormalize on line 41

Thanks @pagrubel, not sure how I missed that one in my edit.  Should be fixed in new commit 00759f39ec1a75a6c4790a05ffe44ceefb6dc5be.

**Review by pagrubel — APPROVED** [review]
Great article, moving this to

**Comment by pagrubel** [comment]
@betterscientificsoftware/bssw-maint This PR is ready to publish, please carry out the finalization steps.

**Review by rinkug — APPROVED** [review]

**Comment by rinkug** [comment]
@karbarz : I published this today since it seems good to me. But pls have a look from your perspective and open another PR,if needed.

**Pull request merged**